source-salesforce-native/source_salesforce_native/resources.py

```diff
@@ -0,0 +1,319 @@
+    enable = details.get('enabled_by_default', False)
+    is_supported_by_bulk_api = details.get('is_supported_by_bulk_api', True)
+    assert isinstance(enable, bool)
+    assert isinstance(is_supported_by_bulk_api, bool)
+    fields = await _fetch_object_fields(log, http, instance_url, name) if should_fetch_fields else None
+
+    if fields:
+        field_soap_types = {fields[field].soapType for field in fields}
+        for soap_type in field_soap_types:
+            if soap_type in SOAP_TYPES_NOT_SUPPORTED_BY_BULK_API:
+                is_supported_by_bulk_api = False
+
+
+    if is_custom_object or cursor_field is not None:
+        return incremental_resource(
+            log,
+            http,
+            config,
+            is_supported_by_bulk_api,
+            bulk_job_manager,
+            rest_query_manager,
+            instance_url,
+            name,
+            fields,
+            enable,
+        )
+    else:
+        return full_refresh_resource(
+            log,
+            http,
+            config,
+            bulk_job_manager,
+            instance_url,
+            name,
+            fields,
+            enable,
+        )
+
+
+# enabled_resources returns resources for only enabled bindings.
+async def enabled_resources(
+    log: Logger, http: HTTPMixin, config: EndpointConfig, bindings: list[common._ResolvableBinding]
+) -> list[common.Resource]:
+    update_oauth_spec(config.is_sandbox)
+    http.token_source = SalesforceTokenSource(oauth_spec=OAUTH2_SPEC, credentials=config.credentials)
+
+    enabled_binding_names: list[str] = []
+
+    for binding in bindings:
+        path: list[str] = binding.resourceConfig.path()
+        enabled_binding_names.append(path[0])
+
+    bulk_job_manager = BulkJobManager(http, log, config.credentials.instance_url)
+    rest_query_manager = RestQueryManager(http, log, config.credentials.instance_url)
+
+    # If we concurrently send multiple requests that exchange the same refresh token for an access token,
+    # some of those requests intermittently fail. 
+    # https://help.salesforce.com/s/articleView?id=xcloud.remoteaccess_oauth_refresh_token_flow.htm&type=5#:~:text=Avoid%20sending%20simultaneous%20requests%20that%20contain%20the%20same%20refresh%20token.%20If%20your%20client%20sends%20identical%20requests%20at%20the%20same%20time%2C%20some%20of%20the%20requests%20fail%20intermittently%20and%20the%20Status%20column%20in%20the%20Login%20History%20displays%20Failed%3A%20Token%20request%20is%20already%20being%20processed.
+    # To avoid this, we make a noop request to set the token_source's access token before using the scatter-gather
+    # technique to make multiple requests concurrently. This prevents the first BUILD_RESOURCE_SEMAPHORE_LIMIT
+    # requests from all exchanging the same access token and encountering that intermittent error.
+    await _fetch_queryable_objects(log, http, config.credentials.instance_url)
```

Comment:
I remember looking at this a little bit in the CDK HTTP module. We should probably add some synchronization there. Doesn't have to be right now though, since what you have here should work.